### ISSUE TITLE
How to get project version and project group for every projects in a multi-projects build, even for subprojects (included by an other project) ?

### ISSUE THREAD

**Issue by Rhamao** [eclipse/buildship#965]
Hi, I am working on the management of multi-project build with Buildship. I've launched an Eclipse Application in debug with `org.eclipse.buildship.core`, where I have 2 projects A and B. A having a dependency on project B : 
`Project A build.gradle` 
![image](https://user-images.githubusercontent.com/47444260/72247298-df211100-35f4-11ea-920b-e9d2c1ceeb41.png)
`Project A settings.gradle`
![image](https://user-images.githubusercontent.com/47444260/72247260-cc0e4100-35f4-11ea-868a-37b3cd5d52ee.png)
`Project B build.gradle`
![image](https://user-images.githubusercontent.com/47444260/72247350-f8c25880-35f4-11ea-9d4a-163fa0fe3be6.png)
`Project B settings.gradle`
![image](https://user-images.githubusercontent.com/47444260/72247399-168fbd80-35f5-11ea-8d5d-4bbcc9067924.png)

To get the publication informations of both projects, I wrote the following script :
```
ImmutableList<IProject> allWorkspaceProjects = CorePlugin.workspaceOperations().getAllProjects();
for (IProject project : allWorkspaceProjects) {
	System.out.println(project + " Opened : " + project.isOpen());
	if(project.isOpen()) {
		GradleBuild build = GradleCore.getWorkspace().getBuild(project).get();
		IProgressMonitor monitor = new NullProgressMonitor();
		ProjectPublications model;
		try {
			model = build.withConnection(connection -> connection.getModel(ProjectPublications.class), monitor);
			System.out.println(model.getProjectIdentifier());
			System.out.println(model.getPublications().getAll());
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}
```
Unfortunately, I can't get identifier of project B. Also, when I look at the  variables of the `build` object ,  it seems that they are always concerning project A, never project B. Even when I do give it an `IProject object` that contains informations about project B, as show in the console screenshot below : 
![image](https://user-images.githubusercontent.com/47444260/72246711-a3397c00-35f3-11ea-8724-13d64517f4cc.png)
I did not figure out why it only gives me informations about project A. Maybe because it is shown as the `DefaultGradlePublication`. However I need this informations. So is there a way to get the version and the group of others projects than project A ❓

**Comment by Rhamao** [comment]
All right. I succeeded in getting B group and versions by getting rid of the dependency between them.

**Comment by Rhamao** [comment]
It seems that something happened when we create a dependency that force the system to recognise A as the only root project.

**Comment by donat** [comment]
The code snippet you posted will ignore the included builds:

    model = build.withConnection(connection -> connection.getModel(ProjectPublications.class), monitor);

You have to use the `ProjectConnection.action(BuildAction)` method.. In the BuildAcgtion implementation you can query the included builds via `BuildController.getIncludedBuilds()`.

**Comment by Rhamao** [comment]
Thank you for your answer. I'll try that. I reopen the issue and rename it.

**Comment by Rhamao** [comment]
I tried to do it as you indicated me. But I can't get get it to work. Here is what I wrote :
```
                   build.withConnection(connection -> {
                        BuildActionExecuter<String> action = connection.action(controller -> {
                            //For some reasons we don't reach what is in brackets there
                            //And we have some errors
                            //controller.getBuildModel().getIncludedBuilds();
                            System.out.println("Do we reach that line ?");
                            return "";
                        });
                        action.forTasks().run();
                        return "";
                    }, monitor);
```
I also get some errors such as : 
```
Caused by: java.lang.ClassNotFoundException: org.eclipse.buildship.core.internal.workspace.GradleClasspathContainerUpdater
...
org.gradle.tooling.BuildException: Could not run build action using Gradle distribution 'https://services.gradle.org/distributions/gradle-6.0-bin.zip'.
```
```
Caused by: org.gradle.api.UncheckedIOException: java.io.InvalidClassException: failed to read class descriptor
...
Caused by: java.lang.ClassNotFoundException: org.eclipse.buildship.core.internal.workspace.GradleClasspathContainerUpdater
```
It'll be great, if you could help getting that working.

**Comment by donat** [comment]
Here's how Buildship queries the `EclipseProject` model for composite projects:
https://github.com/eclipse/buildship/blob/17909a1ddf5bc62f63d15f18d1bdf3de569d0665/org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/CompositeModelQuery.java#L54-L65

Call site:
https://github.com/eclipse/buildship/blob/17909a1ddf5bc62f63d15f18d1bdf3de569d0665/org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/EclipseModelUtils.java#L119-L122

**Comment by donat** [comment]
Please note, that the `BuildAction` class needs to be a separate public class, otherwise Gradle won't be able to load it.

There's another problem you need to be aware of: When you launch a runtime Eclipse instance with PDE then you'll see a different classpath then what's available when the plugin is actually installed. In a production environment, everything is fine, but Gradle won't find the custom BuildAction implementation when launching a runtime eclipse with PDE (i.e. Run As > Eclipse Application). To solve that, you need to load the build action in a separate classloader when PDE is being used. Check out the `IdeFriendlyClassLoading` for the implementation details.

**Comment by Rhamao** [comment]
All right, thank you. I am now able to get the publication and the versions of all projects in a composite build defined with `Settings.includeBuild(java.lang.Object)`. I wrote the following code :
```
System.out.println("    Project " + project.getName());
org.gradle.tooling.model.gradle.GradleBuild toolingBuild = build.withConnection(connection -> connection.getModel(org.gradle.tooling.model.gradle.GradleBuild.class), monitor);
BuildAction<Collection<ProjectPublications>> query = IdeFriendlyClassLoading.loadCompositeModelQuery(ProjectPublications.class);
Collection<ProjectPublications> a = build.withConnection(connection -> {
	return connection.action(query).run();
}, monitor);
for(ProjectPublications pp : a) {
	System.out.println("    " + pp.getPublications());
}
```

**Comment by Rhamao** [comment]
At last, I need to do the same for a composite build defined with `Settings.include(java.lang.String[]) ` instead of `Settings.includeBuild(java.lang.Object)`, since the code I wrote only work for `Settings.includeBuild(java.lang.Object)`. 
So far, I didn't find anything similar that works with `Settings.include(java.lang.String[]) `. It'll be very nice of you, if you could help me one last time :D.

**Comment by donat** [comment]
Sorry for the late reply. 

I'm not sure I'm following. So it works for included builds (i.e. `Settings.includeBuild()`) but not for simple multi-project builds (`Settings.include()`)? 

In any case, here's a thing about the snippet you posted. The `withConnection` method is potentially expensive. I'd restructure the code such that it is only called once.

**Comment by Rhamao** [comment]
Sorry too for late reply, I should have answered sooner :(. 

> I'm not sure I'm following. So it works for included builds (i.e. Settings.includeBuild()) but not for simple multi-project builds (Settings.include())?

Yes that is exactly my problem. I am going back to work on that issue tomorrow. I'll keep you informed of any signifcant progress. However, if you're aware of something that is working for simple multi-project builds (Settings.include()), then I'll be more than pleased to be informed. 

Thanks again for your help.

**Comment by babytang** [comment]
Hi,

We have re-started from the ground and may have found the problem for getting the publications on a simple multi-project build.

We tried by using directly the GradleConnector and instanciating a ProjectConnection to the sub-project on a multi-project builds. This works.

We tried with a BuildAction, and the following works too:
`private void extract(BuildController controller, GradleBuild gradleBuild, ArrayList<String> arrayList) {
        GradleProject project = controller.getModel(GradleProject.class);
        for (GradleProject child : project.getChildren()) {
                ProjectPublications publications = controller.getModel(child, ProjectPublications.class);
                for (GradlePublication pub : publications.getPublications()) {
                        // Here we got the child publications.
                }
        }
}`

In fact we use: `GradleBuild build = GradleCore.getWorkspace().getBuild(project).get();` for getting the build where we can create the ProjectConnection and then get the publications with `getModel(ProjectPublications.class)`.

But what we didn't see is that the instance returned by this method is in fact always the same. This method always returns the GradleBuild associated to the root of the multi-project build.
And so we was always creating a connection to the root and not the expected child.

Is this the wanted behavior ?

**Comment by donat** [comment]
Yes, this is then intended behaviour. `GradleBuild` represents a Gradle build, which is always the same for all projects associated with it. 

You should treat the `withConnection` method as an expensive operation (which it is). So, open one connection per build, recursively iterate through all subprojects/included projects and return the results in one object.

**Comment by Rhamao** [comment]
Thank you. I'll try that.
